### PR TITLE
Use go workspaces to simplify intermodule deps

### LINKED ISSUE #5991: Use go workspaces to simplify dependency versioning
During the SIG meeting on August 10, we agreed that we'd do a proof-of-concept of Go Workspaces to simplify the management of versions (especially replaces statements) in the Go modules for a repository.

I'll work on the PoC, while @bryan-aguilar was requested to help with the tooling, given Bryan's experience with similar tools.

### PR DESCRIPTION
**Pull request by jpkrohling**
I was able to positively confirm that all modules had some effect when added to the go.work. For instance:

```
$ make install-tools
cd ./internal/tools && go install github.com/client9/misspell/cmd/misspell
cd ./internal/tools && go install github.com/golangci/golangci-lint/cmd/golangci-lint
no required module provides package github.com/golangci/golangci-lint/cmd/golangci-lint; to add it:
        go get github.com/golangci/golangci-lint/cmd/golangci-lint
make: *** [Makefile:130: install-tools] Error 1

$ go work use internal/tools/

$ make install-tools
cd ./internal/tools && go install github.com/client9/misspell/cmd/misspell
cd ./internal/tools && go install github.com/golangci/golangci-lint/cmd/golangci-lint
cd ./internal/tools && go install github.com/google/addlicense
cd ./internal/tools && go install github.com/ory/go-acc
cd ./internal/tools && go install github.com/pavius/impi/cmd/impi
cd ./internal/tools && go install github.com/tcnksm/ghr
cd ./internal/tools && go install github.com/wadey/gocovmerge
cd ./internal/tools && go install go.opentelemetry.io/build-tools/checkdoc
cd ./internal/tools && go install go.opentelemetry.io/build-tools/semconvgen
cd ./internal/tools && go install golang.org/x/exp/cmd/apidiff
cd ./internal/tools && go install golang.org/x/tools/cmd/goimports
cd ./internal/tools && go install github.com/jcchavezs/porto/cmd/porto
cd ./internal/tools && go install go.opentelemetry.io/build-tools/multimod
```

Fixes #5991

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
This is still a draft as I'm figuring out what needs to change in terms of CI. Also, I would like to have a word from @bryan-aguilar on what we would need in terms of tooling.

**Comment by bryan-aguilar** [comment]
It's been a little while since I dug into `go.work` files and their uses but I believe I still have a high level understanding. Without knowing more requirements, if we are looking to centralize around a central `go.work` file, then adjustments could be made to the [`crosslink`](https://github.com/open-telemetry/opentelemetry-go-build-tools/tree/main/crosslink) tool to support this. If there are more requirements, multiple `go.work` files for some reason, then I would need to perform a bit more research on implementation details.

As a note, the initial `go.work` [proposal](https://go.googlesource.com/proposal/+/master/design/45713-workspace.md) states that it is not  recommended to check in `go.work` files to version control systems. 
```
go.work files should not be checked into version control repos containing modules so that the go.work file in a module does not end up overriding the configuration a user created themselves outside of the module. The go.work documentation should contain clear warnings about this.
```
Again, It's been a while since I've done digging into the `go.work` files so things may have changed. Is this much different than us automatically inserting replacing statements for common modules? Crosslink already addresses that need.

**Comment by bogdandrutu** [comment]
@jpkrohling @bryan-aguilar while back, @codeboten shared this golang blogpost with me https://go.googlesource.com/proposal/+/master/design/45713-workspace.md#preventing-files-from-being-checked-in-to-repositories

I think their suggestion is different, not sure how do they envision go workspaces to work.

**Comment by jpkrohling** [comment]
> If there are more requirements, multiple go.work files for some reason

That's not the case. I don't think we need that.

> As a note, the initial go.work [proposal](https://go.googlesource.com/proposal/+/master/design/45713-workspace.md) states that it is not recommended to check in go.work files to version control systems.

I wasn't able to find anything authoritative around whether we should or not check in the go.work file, but I found this GitHub issue: https://github.com/golang/go/issues/53502

Apparently, our case is one that would justify checking in the go.work file: a single git repository with multiple modules.

**Comment by jpkrohling** [comment]
> while back, @codeboten shared this golang blogpost with me

That's the proposal @bryan-aguilar linked in his comment.

File: cmd/otelcorecol/go.sum [hunk 1]
```diff
@@ -419,12 +419,16 @@ go.opencensus.io v0.22.3/go.mod h1:yxeiOL68Rb0Xd1ddK5vPZ/oVn4vY4Ynel7k9FzqtOIw=
 go.opencensus.io v0.22.4/go.mod h1:yxeiOL68Rb0Xd1ddK5vPZ/oVn4vY4Ynel7k9FzqtOIw=
 go.opencensus.io v0.23.0 h1:gqCw0LfLxScz8irSi8exQc7fyQ0fKQU/qnC/X8+V/1M=
 go.opencensus.io v0.23.0/go.mod h1:XItmlyltB5F7CS4xOC1DcqMoFqwtC6OG2xF7mCv7P7E=
+go.opentelemetry.io/collector v0.58.0 h1:ofl5qa+vTV69PC9NaZKQjE7MP/49iclDKRppl00WgZg=
+go.opentelemetry.io/collector v0.58.0/go.mod h1:U3TE477WDi3CYhmE7JGinnpIg8qMH1KCBkRmk3BxKyw=
+go.opentelemetry.io/collector/pdata v0.56.0 h1:JD8KjQ7dNZ441xMuVZVu5NRYmkA4vOYGV7w8tkCdyrE=
+go.opentelemetry.io/collector/pdata v0.56.0/go.mod h1:mYcCREWiIJyHss0dbU+GSiz2tmGZ6u09vtfkKTciog4=
+go.opentelemetry.io/collector/semconv v0.56.0 h1:zpQ6IBimBsiVsJibsSM2/13vKtaeteFFIx4bmIiOS6E=
+go.opentelemetry.io/collector/semconv v0.56.0/go.mod h1:EH1wbDvTyqKpKBBpoMIe0KQk2plCcFS66Mo17WtR7CQ=
```
Comment (codeboten) [inline cmd/otelcorecol/go.sum:427]: Should these be here? It looks like its importing dependencies from outside the repo.

File: go.sum [hunk 2]
```diff
@@ -423,6 +423,10 @@ go.opencensus.io v0.22.3/go.mod h1:yxeiOL68Rb0Xd1ddK5vPZ/oVn4vY4Ynel7k9FzqtOIw=
 go.opencensus.io v0.22.4/go.mod h1:yxeiOL68Rb0Xd1ddK5vPZ/oVn4vY4Ynel7k9FzqtOIw=
 go.opencensus.io v0.23.0 h1:gqCw0LfLxScz8irSi8exQc7fyQ0fKQU/qnC/X8+V/1M=
 go.opencensus.io v0.23.0/go.mod h1:XItmlyltB5F7CS4xOC1DcqMoFqwtC6OG2xF7mCv7P7E=
+go.opentelemetry.io/collector/pdata v0.58.0 h1:SKWw4vjd6ZjCuvsCvEzqwBaxvov4YbXnnXkc9C4xMqM=
+go.opentelemetry.io/collector/pdata v0.58.0/go.mod h1:iMv7Pz+hRthi30rkYkwLVusxQ94GU4pPJgFq7gjGcBk=
+go.opentelemetry.io/collector/semconv v0.58.0 h1:wk9KXVnt8IRdNzD9mmdW3d1M/IJ3HyLp1Lz2ZY1fBCM=
+go.opentelemetry.io/collector/semconv v0.58.0/go.mod h1:aRkHuJ/OshtDFYluKEtnG5nkKTsy1HZuvZVHmakx+Vo=
```
Comment (codeboten) [inline go.sum:429]: Same question here

File: cmd/otelcorecol/go.sum (hunk 1 above)
Comment (jpkrohling) [reply cmd/otelcorecol/go.sum]: Not quite sure what happened. Perhaps the sum was built while one of the modules still wasn't part of the workspace? In any case, a rebase helped.

**Pull request closed without merging**